extensions/interactions/GraphInput/GraphInput.js

```diff
@@ -187,362 +191,476 @@ oppia.directive('graphViz', function() {
+      });
 
+      if (pivot === null) {
+        return null;
+      }
 
-      // Styling functions
-      var DELETE_COLOR = 'red';
-      var HOVER_COLOR = 'aqua';
-      var SELECT_COLOR = 'orange';
-      var DEFAULT_COLOR = 'black';
-      $scope.getEdgeColor = function(index) {
-        if ($scope.state.currentMode === _MODES.DELETE &&
-            index === $scope.state.hoveredEdge &&
-            $scope.canDeleteEdge) {
-          return DELETE_COLOR;
-        } else if (index === $scope.state.hoveredEdge) {
-          return HOVER_COLOR;
-        } else if ($scope.state.selectedEdge === index) {
-          return SELECT_COLOR;
-        } else {
-          return DEFAULT_COLOR;
-        }
-      };
-      $scope.getVertexColor = function(index) {
-        if ($scope.state.currentMode === _MODES.DELETE &&
-            index === $scope.state.hoveredVertex &&
-            $scope.canDeleteVertex) {
-          return DELETE_COLOR;
-        } else if (index === $scope.state.currentlyDraggedVertex) {
-          return HOVER_COLOR;
-        } else if (index === $scope.state.hoveredVertex) {
-          return HOVER_COLOR;
-        } else if ($scope.state.selectedVertex === index) {
-          return SELECT_COLOR;
-        } else {
-          return DEFAULT_COLOR;
+      // swap the pivot and successor and reverse the suffix
+      var tmp = permutation[pivot];
+      permutation[pivot] = permutation[successor];
+      permutation[successor] = tmp;
+      permutation = permutation.concat(permutation.splice(pivot + 1).reverse());
+      return permutation;
+    },
+    areAdjacencyMatricesEqualWithPermutation: function(adj1, adj2, permutation) {
+      var numVertices = adj1.length;
+      for (var i = 0; i < numVertices; i++) {
+        for (var j = 0; j < numVertices; j++) {
+          if (adj1[permutation[i]][permutation[j]] != adj2[i][j]) {
+            return false;
+          }
         }
-      };
-      $scope.getDirectedEdgeArrowPoints = function(index) {
-        return graphDetailService.getDirectedEdgeArrowPoints($scope.graph, index);
-      };
-      $scope.getEdgeCentre = function(index) {
-        return graphDetailService.getEdgeCentre($scope.graph, index);
-      };
-    }]
-  }
-});
+      }
+      return true;
+    }
+  };
+}]);
+
+oppia.factory('graphInputRulesService', [
+    'graphUtilsService', function(graphUtilsService) {
+  var IsIsomorphic = function(graph1, graph2) {
+    if (graph1.vertices.length != graph2.vertices.length) {
+      return false;
+    }
+
+    var adj1 = graphUtilsService.constructAdjacencyMatrix(graph1);
+    var adj2 = graphUtilsService.constructAdjacencyMatrix(graph2);
+
+    // Check against every permutation of vectices.
+    var numVertices = graph2.vertices.length;
+    var permutation = [];
+    for (var i = 0; i < numVertices; i++) {
+      permutation.push(i);
+    }
+    while (permutation !== null) {
+      var foundIsomorphism = true;
+      if ((!(graph1.isLabeled || graph2.isLabeled) ||
```

Comment:
hmm, this is hard to read. Break it out?

var doLabelsMatch = ...
if (doLabelsMatch && areAdjacencyMatricesEqual(...)) {
  return true;
}